src/base/utils/tar.cpp

```diff
@@ -0,0 +1,134 @@
+/*
+ * Bittorrent Client using Qt and libtorrent.
+ * Copyright (C) 2020  sledgehammer999 <hammered999@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "tar.h"
+
+#include <cstring>
+
+#include <QByteArray>
+#include <QHash>
+#include <QString>
+
+namespace
+{
+    const int TAR_HEADER_SIZE = 512;
+
+    bool tarEOF(const char *data)
+    {
+        // static variables are auto initialized to zero
+        static char testBlock[2 * TAR_HEADER_SIZE];
```

Comment:
The purpose was to be efficient in initializing/allocating memory.
This is what is alluded here: https://stackoverflow.com/questions/6938219/how-to-check-whether-all-bytes-in-a-memory-block-are-zero

If the above doesn't really matter, then should I do this?
```c++
char testBlock[2 * TAR_HEADER_SIZE] = {};
```

Reply:
Edited previous comment.

Reply:
IMO, it's not so good idea to copy some memory just to check for buffer end. Why not perform integer calculations instead?

Reply:
The other way I see it, is just loop in the buffer and check each byte for zero. But the SO answers say that his might not be optimal/efficient.

If there is another way, please give an example.

Reply:
Oops... It's `memcmp`! Forget my previous comment.

Reply:
>If the above doesn't really matter, then should I do this?

Yeah. 
If using QbyteArray you can use this constructor QByteArray::QByteArray(int size, char ch) and compare them with QByteArray::startsWith(const QByteArray &ba).

Reply:
I think I have a better alternative to test for EOF (and possibly gracefully handle garbage archives/data):

As mentioned, a tar archive looks like this `header+filedata+header+filedata+...+empty header+empty header`
Each time we are finished with a file we are expecting either a normal header or at least 2 empty ones. And the first field on a valid header is the filename field.
So why not check if the 1st byte in the header is null, and if yes treat it as EOF?

Reply:
>As mentioned, a tar archive looks like this `header+filedata+header+filedata+...+empty header+empty header`

Very strange way of indicating the end of the file. It is clear that the "real" end of the file could not be used, because it was designed for streaming media, but why two empty headers and not one? And what really about not empty header but just invalid?

Reply:
>So why not check if the 1st byte in the header is null, and if yes treat it as EOF?

At least we can use simplified algorithm.
1. If next byte is not null try to load header and so on...
2. If next byte is null stop processing (but check whether there's valid EOF to distinguish between correct/broken files).

Reply:
>If next byte is not null try to load header and so on...

Even if next byte is not null, we can't really extract a filename from that. There isn't supposed to be padding* between filedata and the next header.

*padding exists at the end of filedata to pad to full 512bytes block.

>but check whether there's valid EOF to distinguish between correct/broken files

Hmm, this check needs to done inside `TarFile::fileData()`. But a variation of it. Not detect null bytes but actually test that there is enough space left in the buffer for the filesize indicated by the header.

Reply:
>Even if next byte is not null, we can't really extract a filename from that. 

I mean first byte of (possible) next header.
E.g.:
1. Try to load header.
2. If there isn't enough data - error.
3. Perform the "optimize" check for null in first byte of header.
4. If it is null, check for correct EOF at this position. If there's no valid EOF - error.
5. If it isn't null try to parse header and perform other checks on it. If header is invalid - error.
6. Try to load file.
7. If there isn't enough data - error.
8. Perform some checks on file data (e.g. checksum comparison). If it fails - error.
9. Try to shift to next header (i.e. skip padding). If there's no enough data - error.
10. Start from beginning.

Have I missed something? I didn't read full format specs.